benches/engines/mod.rs

```diff
@@ -0,0 +1,232 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate criterion;
+extern crate kvproto;
+extern crate test_util;
+extern crate tikv;
+
+use criterion::{Bencher, Criterion};
+use kvproto::kvrpcpb::Context;
+
+use test_util::generate_random_kvs;
+use tikv::storage::engine::{BTreeEngine, Engine, Modify, Snapshot, TestEngineBuilder};
+use tikv::storage::{Key, CF_DEFAULT};
+
+const DEFAULT_KEY_LENGTH: usize = 64;
+const DEFAULT_ITERATIONS: usize = 1000;
+
+#[derive(Debug, Clone)]
+enum EngineType {
+    BTreeEngine,
+    RocksDB,
+}
+
+fn fill_engine_with<E: Engine>(engine: &E, expect_engine_keys_count: usize, value_length: usize) {
+    let mut modifies: Vec<Modify> = vec![];
+    if expect_engine_keys_count > 0 {
+        let kvs = generate_random_kvs(expect_engine_keys_count, DEFAULT_KEY_LENGTH, value_length);
+        for (key, value) in kvs {
+            modifies.push(Modify::Put(CF_DEFAULT, Key::from_raw(&key), value))
+        }
+    }
+    let ctx = Context::new();
+    let _ = engine.async_write(&ctx, modifies, Box::new(move |(_, _)| {}));
+}
+
+/// Measuring the performance of Engine::snapshot()
+fn engine_snapshot_bench<E: Engine>(
+    bencher: &mut Bencher,
+    engine: &E,
+    engine_keys_count: usize,
+    value_length: usize,
+) {
+    fill_engine_with(engine, engine_keys_count, value_length);
+    let ctx = Context::new();
+
+    bencher.iter(|| {
+        engine.snapshot(&ctx).is_ok();
+    })
+}
+
+/// Actually, it measures the performance of Snapshot::get(), skipping the Engine::snapshot();
+fn engine_get_bench<E: Engine>(
+    bencher: &mut Bencher,
+    engine: &E,
+    iterations: usize,
+    engine_keys_count: usize,
+    value_length: usize,
+) {
+    fill_engine_with(engine, engine_keys_count, value_length);
+    let ctx = Context::new();
+    let test_kvs = generate_random_kvs(iterations, DEFAULT_KEY_LENGTH, value_length);
+
+    bencher.iter_with_setup(
+        || engine.snapshot(&ctx).unwrap(),
+        |snap| {
+            for (key, _) in &test_kvs {
+                snap.get(&Key::from_raw(key)).is_ok();
+            }
+        },
+    );
+}
+
+/// Measure the performance of Engine::put()
+fn engine_put_bench<E: Engine>(
+    bencher: &mut Bencher,
+    engine: &E,
+    write_count: usize,
+    value_length: usize,
+) {
+    let test_kvs = generate_random_kvs(write_count, DEFAULT_KEY_LENGTH, value_length);
+    let ctx = Context::new();
+    bencher.iter(|| {
+        for (key, value) in &test_kvs {
+            let _ = engine.put(&ctx, Key::from_raw(&key), value.clone());
```

Comment:
1. `Key::from_raw(..)` and `value.clone()` can be moved to setup, so that their cost won't be counted.
2. Wrap with `black_box`. For example, in real applications, the value of `Context` is unknown (which is passed in from requests), so in benchmarks, we also need to make sure that they are non-constant, i.e. wrap with black_box, so that compiler won't optimize `engine.put` specifically for this kind of constant value.